### DISCUSSION TITLE
Letting Embedded Boundary work with AMRcore model mainly in Fortran interface

### DISCUSSION THREAD

**Discussion by paradoxknight1** [AMReX-Codes/amrex#4248] [Q&A]
Dear community,

I'm currently working with the fortran interface amrcore model. And there is a requirement of implement the Embedded Boundary function.
Based on the docuemnt, it seems that to build a Embedded Boundary, the Geometry object is needed. However, through the usage of amrcore model, the geom, BoxArrays, and DistributionMapping are both automatically building through the C++ part of the AMRmesh model.
The thing I would like to ask is that if it is possible to build the embedded boundary in the Fortran's  inherit virtual functions.

What's more, as it mentioned in https://github.com/AMReX-Codes/amrex/discussions/3645.
"Usually one calls EB2:Build with the Geometry at the max level at the beginning." What exactly is the max level? As the Geometry in Amrcore is build automatically.
Is it possible to build a geometry seperately aginest the Amrcore framework?

Many thanks.

**Comment by WeiqunZhang** [comment]
If you are willing to write some codes in C++, yes you can build Embedded Boundary in a Fortran code.

Re: your questions about max level and Geometry, let's suppose you are using `amrex_amrcore_module`. That module has `type(c_ptr) :: amrcore`. On the Fortran side, you can call a C++ function (written by you) and pass that pointer to the C++ side. That pointer is `amrex::AmrCore*`. Then you have everything you need regarding max level, Geometry, etc.

**Reply by paradoxknight1** [reply]
Thanks for your timely reply. Some code in C++ is fine for me.
Just to clarify my understanding, that is to say, the amr-core module on the C++ side only builds a geometry distribution map once. Therefore, what I need to do is wait till the initialization process is finished and then apply the embedded boundary module to modify the automatically built geometry object.
May I check under that condition the main function should be something like
![image](https://github.com/user-attachments/assets/e7877f8a-58fc-44f6-9ca1-4b494225d91f)

Many Thanks for your kind help again.

**Reply by WeiqunZhang** [reply]
I think you want to initialize EB after amrex_amrcore_init but before my_amr_init so that my_amr_init can use EB.

**Reply by paradoxknight1** [reply]
Thanks for your help. 
Here is another problem from the C++ and Fortran Interface: To fill the cut cell's value about the embedded boundary, a huge amount of information from the EB module is needed.  Apart from EBcellFlagFab, which could be transferred to the Fortran Interface as an integer matrix, **volume fraction, and normal vector** are both needed. However, they are MultiCut objects, which are not defined on the Fortran side.
There are two options from my perspective,
1. Transfer the MFI object to C++, which builds a CutFab object, which could then be transferred as a Real matrix to Fortran.
2. If the MultiCutFab object could be converted as a MultiFab tand hen converted to the Fortran side? Since it only has data on boxes that contain cut cells,  just need to handle it carefully.

Many thanks for your kind help.

**Reply by paradoxknight1** [reply]
May I check which option is available.
For the first option, I'm not sure if the Mfiter object could be transferred directly from C++ to Fortran, as their definition might be different.
For the second option, I'm not sure if it is OK for the transfer from MultiCutFab to MultiFab.

**Reply by WeiqunZhang** [reply]
`amrex_mfiter` in Fortran is
```
  type, public :: amrex_mfiter
     type(c_ptr)      :: p       = c_null_ptr
     ...     
```
That `p` is a C pointer to `amrex::MFIter`.

MultiCutFab has a member function `ToMultiFab` you can use. For example
```
MultiFab centroid = ebfactory.getCentroid().ToMultiFab();
```

**Reply by WeiqunZhang** [reply]
To get that MultiFab to the Fortran side, you need an object on the heap. So you probably need to do something like
```
MultiFab* p_centroid_mf = new MultiFab(ebfactory.getCentroid().ToMultiFab());
```
